src/st.py

```diff
@@ -1,14 +1,134 @@
+from __future__ import annotations
 import argparse
+from dataclasses import dataclass
+import search
 
+@dataclass
+class Node:
+    start: int
+    end: int
+    child: Node | None # If end==len(x) then child is is the index
+    sib : Node | None
+
+    def __str__(self):
+        return f"{self.start},{self.end},{type(self.child)}, {type(self.sib)}"
+    
+    def __len__(self):
+        return self.end-self.start
+
+SENTINEL = "$"
+
+def getAllSuffixes(x):
+    for i in range(1, len(x)):
+        yield x[i:]+SENTINEL
+    yield SENTINEL
```

Comment:
I am not super happy about this. This creates all the suffixes in $O(n^2)$ time, and if you don't free them (or let the GC do it) also space. It is entirely unnecessary as the index `i` always defines the suffix `x[i:]`. There is no need for copying.